dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/AssumeSolver.java

```diff
@@ -4,34 +4,60 @@
 import com.dat3m.dartagnan.encoding.PropertyEncoder;
 import com.dat3m.dartagnan.encoding.SymmetryEncoder;
 import com.dat3m.dartagnan.encoding.WmmEncoder;
+import com.dat3m.dartagnan.program.Program;
 import com.dat3m.dartagnan.utils.Result;
+import com.dat3m.dartagnan.verification.Context;
 import com.dat3m.dartagnan.verification.VerificationTask;
+import com.dat3m.dartagnan.wmm.Wmm;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.sosy_lab.common.configuration.Configuration;
 import org.sosy_lab.common.configuration.InvalidConfigurationException;
 import org.sosy_lab.java_smt.api.*;
 
 import static com.dat3m.dartagnan.utils.Result.FAIL;
 import static com.dat3m.dartagnan.utils.Result.PASS;
 import static java.util.Collections.singletonList;
 
-public class AssumeSolver {
+public class AssumeSolver extends ModelChecker {
 
     private static final Logger logger = LogManager.getLogger(AssumeSolver.class);
 
+    private final SolverContext ctx;
+    private final ProverEnvironment prover;
+    private final VerificationTask task;
```

Comment:
All solvers have these. Can we simply move them to `ModelChecker` as `protected`? The class `TwoSolvers` have two `ProverEnviroments`, but I guess it can use the one in `ModelChecker` as one and just add a second one in that class.

Reply:
I wouldn't do this right now. I think the whole idea of explicitly providing the `ProverEnvironment` and `SolverContext` as input is a little outdated.
We only do this to be able to access the solver model to generate some output (e.g. check the type of violation).
However, now that we have actual `ModelChecker` instances, it is reasonable that these manage their own `ProverEnviroment/SolverContext` and just make a `Model` accessible, similar to how the SMT solver works: On `FAIL`, we can call `modelChecker.getCounterexample` (or `modelChecker.getModel`). Ideally, we generalize the `ExecutionModel` class to extract complete models (including violated assertions/properties as well as local events) so that we can return those instead of SMT-solver models.
That is what I meant by saying
> making more use of the fact that we can instantiate solvers now
All this requires quite a bit of rework though.

For the time being, I think it is fine to have those variables in each class separately. We will need to think about what exactly we want to achieve with the `ModelChecker` base class (surely not just sharing 3 variables!), and once we decided, we can refactor if necessary.